core/src/main/java/org/elasticsearch/search/slice/TermsSliceQuery.java

```diff
@@ -0,0 +1,140 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.slice;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.index.PostingsEnum;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.DocIdSet;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.DocIdSetBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * A Query that builds a DocIdSet per segment to slice the documents.
+ */
+final class TermsSliceQuery extends SliceQuery {
+    private Object coreCacheKey;
+    private DocIdSet[] disis;
+
+    public TermsSliceQuery(Object coreCacheKey, Slice slice) {
+        super(slice);
+        Objects.requireNonNull(coreCacheKey);
+        Objects.requireNonNull(slice);
+        this.coreCacheKey = coreCacheKey;
+        this.disis = null;
+    }
+
+    @Override
+    public Query rewrite(IndexReader reader) throws IOException {
+        return this;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other instanceof TermsSliceQuery) {
+            return super.equals(other) &&
+                coreCacheKey.equals(((TermsSliceQuery) other).coreCacheKey);
+        }
+        return false;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), coreCacheKey);
+    }
+
+    @Override
+    public String toString(String field) {
+        return "TermSliceQuery [field=" + getSlice().getField() +
+            ", id=" + getSlice().getId() +
+            ", max=" + getSlice().getMax() +
+            ", coreCacheKey=" + coreCacheKey + "]";
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, boolean needsScores) throws IOException {
+        // If the IndexReader has not changed it is safe to reuse the DocIdSets.
+        if (searcher.getIndexReader().getCoreCacheKey().equals(coreCacheKey) == false) {
+            throw new IOException("IndexReader has changed, you must recreate the query with the new searcher.");
+        }
+        if (disis == null) {
+            disis = TermsSliceQuery.build(searcher.getIndexReader(), getSlice());
+        }
+        return new ConstantScoreWeight(this) {
+            @Override
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+                if (disis[context.ord] == null) {
```

Comment:
as I mentioned above - can we just calculate the bitset for the one slice here and let it be cached?